blockchain/blockchain.go

```diff
@@ -570,6 +581,27 @@ func (b *Blockchain) SanityCheckNewHeight(block *core.Block, stateUpdate *core.S
 		return err
 	}
 
+	if head, err := b.L1Head(); err == nil {
+		var header *core.Header
+		if head.BlockNumber == block.Number {
+			header = block.Header
+		} else {
+			header, err = b.BlockHeaderByNumber(head.BlockNumber)
+		}
+
+		if err == nil && !head.BlockHash.Equal(header.Hash) {
+			return ErrBlockConflictsWithL1{
+				L1HeadHash:  head.BlockHash,
+				L2BlockHash: header.Hash,
+				Height:      head.BlockNumber,
+			}
+		} else if !errors.Is(err, db.ErrKeyNotFound) {
+			return err
+		}
+	} else if !errors.Is(err, db.ErrKeyNotFound) {
+		return err
+	}
+
```

Comment:
This should be done at the time of Store operation to be on the safe side. `SanityCheckNewHeight` is called async to the store&L1 verification process, so it might create race conditions.

Reply:
How I think it should work is;

- Before storing a new L2 block, make sure it agrees with the L1 head. If not, reject
- After storing a L1 head, make sure the L2 chain agrees with the new L1 head. If not, revert the L2 chain as many blocks as necessary.

Reply:
> After storing a L1 head, make sure the L2 chain agrees with the new L1 head.  If not, revert the L2 chain as many blocks as necessary.

This one is tough with the current design. Rolling back the L2 chain in `SetL1Head` creates a race between the synchronizer and L1 verifier (please correct me if I'm wrong). At least it is tough to reason about.

The current implementation will eventually resolve any discrepancy between L1 and L2 in `SanityCheckNewHeight` (will move that to `Store` per your first comment).

Does that seem sufficient?

Reply:
> Rolling back the L2 chain in SetL1Head creates a race between the synchronizer and L1 verifier (please correct me if I'm wrong).

Not really. We only allow a single write transaction, `SetL1Head` blocks all other `Store` calls. So if `SetL1Head` ends up reverting the L2 chain, next call to `Store` that synchronizer makes will fail and synchronizer will have to restart the sync loop and try to sync new blocks on the new L1 verified chain.

Reply:
Sounds good. I'll add a check to `SetL1Head`.

> We only allow a single write transaction

Do we specify this somewhere, or is it a Pebble thing?

Reply:
> Do we specify this somewhere, or is it a Pebble thing?

It is something that I thought would help us reason with concurrency  

https://github.com/NethermindEth/juno/blob/8bb4c7e6e6f2dcddcb22b0b1378866fc624d9c71/db/pebble/db.go#L53